### ISSUE TITLE
support function instance scoped DI

### ISSUE THREAD

**Issue by ragogitpub** [introtocomputerscience/azure-function-autofac-dependency-injection#16]
**Is your feature request related to a problem? Please describe.**
Project supports function level configs. but there isnt a way to do instance specific scoped services. for eg if instead of TraceWriter an ILogger instance had to be injected.

**Describe the solution you'd like**
For the function itself as well as the other classes that are involved in processing a particular function execution, there should be a way to register a class so that for a given function invocation, the same class instance is returned. In other words the lifetime of this type of object should be for the function invocation. "function request scope". Would this be possible ?

**Comment by rawrspace** [comment]
Can you clarify for me what you want to do? Right now you specify the config at the class level and the inject keyword works for the function. You can simply pass the injected object to other functions with that function invocation. Are you wanting the object to exist between multiple function invocations like a singleton?

Thanks,
CJ

**Comment by ragogitpub** [comment]
i think Boris is describing it here https://blog.wille-zone.de/post/azure-functions-proper-dependency-injection/

basically i might have services that are

1. singleton across all function invocations (appsettings type stuff for instance)
2. singleton for a particular function (myService type stuff)
3. singleton for a particular function invocation instance. meaning any binding that is in context of a particular instance call (by functionInstanceId) must get this singleton. For eg TraceWriter type stuff.
4. transient meaning each time a bind is called there is a new instance is returned.

**Comment by AdrianPell** [comment]
I've been taking a look at this today - basically because it started to bite me too! - and I have something of a POC for this, but I have some questions since I don't think I have exactly what was being requested.

The current code instantiates one Autofac scope _per class_.  That is, every class that is decorated with a DependencyInjectionConfig attribute is its own scope.  There is no global scope across all classes, nor is there per-function or per-function-invocation scope.  This means that every non-transient object gets added to the per-class root scope and lives until the function app is terminated.  Transient bindings gets generated afresh every time - even if they are included multiple times in a chain of constructors. In anything but a simple function app, this starts to have problems - hence the original question.

Now, back to the earlier proposal …

The current design permits a different DI configuration to be specified for each class.  This provides great flexibility, but doesn't easily permit:
- functions defined in different classes to call each other, even if they specify the same config as a new config (and a new root) will be created for each class.  Not sure if using different classes is good practice (it works), or whether partial classes would be a better way to split large function sets (e.g. where durable functions are involved).
- singletons at the app level as, again, there is no good way to specify the DI config across classes

I have a prototype that permits:
1. singletons at the class level - specified as SingleInstance()
2. per-function-invocation instances that propagate down the constructor call tree and are disposed when the function completes or errors out - specified as InstancePerLifetimeScope()
3. transient instances that are created each time they are encountered

I think it would be straightforward to introduce a per-function instance as well - it might have be specified as InstancePerMatchingLifetimeScope("function") or something like that.

Thoughts? Is this worth pursuing?

**Comment by rawrspace** [comment]
I am definitely interested in the functionality and I have worked on a prototype before but I got caught up with work and haven't been able to continue pursuing it. If you would like to take a stab at it I am always looking for contributors. Just make sure that whatever functionality you introduce it does not break what already exists.

If you get into and run into any issues let me know and I can have a look with you to see if a fresh set of eyes can help.

Thanks @AdrianPell
